receiver/mongodbatlasreceiver/alerts.go

```diff
@@ -289,6 +520,86 @@ func payloadToLogs(now time.Time, payload []byte) (plog.Logs, error) {
 	return logs, nil
 }
 
+// alertRecord wraps a sync Map so it is goroutine safe as well as
+// can have custom marshaling
+type alertRecord struct {
+	sync.Mutex
+	LastRecordedTime *time.Time `mapstructure:"last_recorded"`
+}
+
+func (a *alertRecord) SetLastRecorded(lastUpdated *time.Time) {
+	a.Lock()
+	a.LastRecordedTime = lastUpdated
+	a.Unlock()
+}
+
+func (a *alertsReceiver) syncPersistence(ctx context.Context) error {
+	if a.storageClient == nil {
+		return nil
+	}
+	c := *a.storageClient
+	cBytes, err := c.Get(ctx, alertCacheKey)
+	if err != nil {
+		a.record = &alertRecord{}
+		return err
+	}
+
+	if cBytes != nil {
+		var cache alertRecord
+		if err = json.Unmarshal(cBytes, &cache); err != nil {
+			return fmt.Errorf("unable to decode stored cache: %w", err)
+		}
+		a.record = &cache
+		return nil
+	}
+
+	// create if not already loaded
+	newC := &alertRecord{}
+	a.record = newC
+
+	return nil
+}
+
+func (a *alertsReceiver) writeCheckpoint(ctx context.Context) error {
+	if a.storageClient == nil {
+		return nil
+	}
+	c := *a.storageClient
+	marshalBytes, err := json.Marshal(&a.record)
+	if err != nil {
+		return fmt.Errorf("unable to write checkpoint: %w", err)
+	}
+	return c.Set(ctx, alertCacheKey, marshalBytes)
+}
+
+func (a *alertsReceiver) applyFilters(pConf ProjectConfig, alerts []mongodbatlas.Alert) []mongodbatlas.Alert {
+	// default behavior, don't exclude or include projects
+	if len(a.projectClusterInclusions) == 0 && len(a.projectClusterExclusions) == 0 {
+		return alerts
+	}
+	filtered := []mongodbatlas.Alert{}
+	for _, alert := range alerts {
+		// if exclusions are specified, exclude them
+		if len(a.projectClusterExclusions) > 0 {
+			if exclude, ok := a.projectClusterExclusions[projectClusterKey(pConf, alert.ClusterName)]; exclude && ok {
```

Comment:
I ended up spinning up the concept of a `max_alert_processing` which is defaulted to 50 and iterating over the arrays. If we feel like we need to optimize for the lookup further we can maybe make a follow-up PR? That way we can potentially unify how the current logs solution and alerts solution do it rather than further expanding the scope of this PR?

Reply:
I'm confused by your response. My feedback here is for a way to simplify alert filtering by cluster.

Reply:
I guess this is what I did 


```go
func (a *alertsReceiver) applyFilters(pConf ProjectConfig, alerts []mongodbatlas.Alert) []mongodbatlas.Alert {
        ...
	for _, alert := range alerts {
                ...
		if len(pConf.ExcludeClusters) > 0 {
			var shouldExclude = false
			for _, exclusion := range pConf.ExcludeClusters {
				if alert.ClusterName == exclusion {
					shouldExclude = true
				}
			}
			if shouldExclude {
				continue
			}
		}

		if len(pConf.IncludeClusters) > 0 {
			var shouldInclude = false
			for _, inclusion := range pConf.IncludeClusters {
				if alert.ClusterName == inclusion {
					shouldInclude = true
				}
			}
			if !shouldInclude {
				continue
			}
		}

		filtered = append(filtered, alert)
               ...
	}
           ...
	return filtered
}
```

The logs portion of this receiver also uses this struct for filtering albeit in the scope of actually grabbing a cluster from the API rather than working from Project Alerts.

https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/c5cd3a2bbe34fff6e6499052a214a38333856c0d/receiver/mongodbatlasreceiver/logs.go#L148-L171

I think there's probably some unification that can be done separate from this PR and more isolated is what I'm suggesting.

Reply:
```go
for _, alert := range alerts {
	...
	for _, exclusion := range pConf.ExcludeClusters {
		...
```

Here's the problem. We have a time complexity that is unnecessarily bad here. I'm not one to obsess over optimization, but this is a basic "when to use a slice vs when to use a map" type of situation. I'm not asking for anything complicated, just that you use a map to check the cluster because it is the appropriate structure here.

I get that you don't want to load up the `ProjectConfig` struct with fields that are specific to alerts, and I would generally agree here. However, this PR already does exactly that kind of thing in both the config struct and the receiver struct. And, the `ProjectConfig` struct already has a field that is only applicable elsewhere. I agree that a future PR could do a lot to separate concerns though.